# エラー

[Sumor Cloud](https://sumor.cloud) ツール。  
[さらに詳しく](https://sumor.cloud/error) エラーハンドリングライブラリは、複数言語、事前定義のエラーコード、データの受け渡しをサポートしています。

[![CI](https://github.com/sumor-cloud/error/actions/workflows/ci.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ci.yml)
[![Test](https://github.com/sumor-cloud/error/actions/workflows/ut.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ut.yml)
[![Coverage](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml)
[![Audit](https://github.com/sumor-cloud/error/actions/workflows/audit.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/audit.yml)

## インストール

```bash
npm i @sumor/error --save
```

## 前提条件

### Node.JS バージョン

Node.JS バージョン 16.x 以上が必要です

### Node.JS ES モジュールが必要

このパッケージはESモジュールで書かれているため、`package.json` ファイル内の以下のコードを変更してください:

```json
{
  "type": "module"
}
```

## 使用法

### エラーの定義

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    USER_NOT_FOUND: 'ユーザーが見つかりません',
    USER_EXISTED: 'ユーザー {name} は既に存在しています'
  }
})

throw new MyError('USER_NOT_FOUND')
// output: Error: ユーザーが見つかりません

throw new MyError('USER_EXISTED', { name: 'Alice' })
// output: Error: ユーザー Alice は既に存在しています
```

### 複数言語

```js
import defineError from '@sumor/error'

const MyError = defineError({
  language: 'en', // デフォルトの言語
  code: {
    USER_NOT_FOUND: 'ユーザーが見つかりません',
    USER_EXISTED: 'ユーザー {name} は既に存在しています'
  },
  i18n: {
    zh: {
      USER_NOT_FOUND: 'ユーザーが見つかりません',
      USER_EXISTED: 'ユーザー {name} は既に存在しています'
    }
  }
})

const error = new MyError('USER_EXISTED', { name: 'Alice' })
error.language = 'en' // エラー言語を変更
console.log(error)
// output: Error: ユーザー Alice は既に存在しています

error.language = 'zh' // エラー言語を変更
console.log(error)
// output: Error: ユーザー Alice は既に存在しています
```

### エラーをJSONに変換

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    USER_NOT_FOUND: 'ユーザーが見つかりません',
    USER_EXISTED: 'ユーザー {name} は既に存在しています'
  }
})

const error = new MyError('USER_EXISTED', { name: 'Alice' })
console.log(error.json())
// output: {"code":"USER_EXISTED","message":"ユーザー Alice は既に存在しています"}
```

### 基本となるエラー

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    FIELD_VERIFY_FAILED: 'フィールドの検証に失敗しました',
    FIELD_CANNOT_EMPTY: 'フィールド {name} は空にできません',
    FIELD_TOO_LONG: 'フィールド {name} は長すぎます'
  },
  i18n: {
    zh: {
      FIELD_VERIFY_FAILED: 'フィールドの検証に失敗しました',
      FIELD_CANNOT_EMPTY: 'フィールド {name} は空にできません',
      FIELD_TOO_LONG: 'フィールド {name} は長すぎます'
    }
  }
})

const error = new MyError('FIELD_VERIFY_FAILED', {}, [
  new MyError('FIELD_CANNOT_EMPTY', { name: 'ユーザー名' }),
  new MyError('FIELD_TOO_LONG', { name: 'パスワード' })
])

console.log(error.json())
/* 
output: 
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"フィールドの検証に失敗しました",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"フィールド ユーザー名 は空にできません"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"フィールド パスワード は長すぎます"
    }
  ]
}
*/

error.language = 'zh'
console.log(error.json())
/*
output:
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"フィールドの検証に失敗しました",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"フィールド ユーザー名 は空にできません"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"フィールド パスワード は長すぎます"
    }
  ]
}
*/
```

### 標準エラーを組み合わせる

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    FIELD_VERIFY_FAILED: 'フィールドの検証に失敗しました',
    FIELD_CANNOT_EMPTY: 'フィールド {name} は空にできません',
    FIELD_TOO_LONG: 'フィールド {name} は長すぎます'
  }
})

const error = new MyError('FIELD_VERIFY_FAILED', {}, [
  new MyError('FIELD_CANNOT_EMPTY', { name: 'ユーザー名' }),
  new MyError('FIELD_TOO_LONG', { name: 'パスワード' }),
  new Error('不明なエラー')
])

console.log(error.json())
/*
output:
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"フィールドの検証に失敗しました",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"フィールド ユーザー名 は空にできません"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"フィールド パスワード は長すぎます"
    },{
      "code":"UNKNOWN_ERROR",
      "message":"不明なエラー"
    }
  ]
}
 */
```
